src/dataflow-types/src/client.rs

```diff
@@ -1143,6 +1369,177 @@ pub mod tcp {
         )
     }
 
+    /// The server side gRPC implementation that will run in computed.
+    ///
+    /// There are two main tasks involved: The gRPC callback implementations will execute in their own tasks,
+    /// receive commands from the network and send responses to the network. Upon reception of a command,
+    /// the implementation will put it in a mpsc queue, out of which the consumer (running in another
+    /// task) will read it with a recv call. The same goes for the send path: The consumer calls `send` which
+    /// puts the response in a mpsc queue, from which the gRPC stubs will read and send it over the network.
+    ///
+    /// This is the shared datastructure between server and consumer.
+    pub struct GrpcComputedShared {
+        // These are endpoints for the consumer. The other end of these queues is consumed by the
+        // stream implementation. `queue` is None if no client is connected, otherwise
+        // the endpoints are forwarded to the single client. If a client is connecting but a client
+        // is already connected, this mutex is used to block. If the consumer calls send or recv and queue
+        // is None, the consumer will wait on the queue_change notification to proceed only when a
+        // client has connected.
+        queue: Mutex<
```

Comment:
why do we need this mutex? I don't see anywhere we need `&mut` access from a `GrpcComputedShared`

Reply:
maybe we need it for the receiver side, but we should just have the mutex on the smallest scope possible

Reply:
ah I forgot about `recv` needing `&mut`